### PR TITLE
Allow multiple includes of Dry::Configurable

### PR DESCRIPTION
**Pull request by timriley**
Instead of raising AlreadyIncludedError (now removed), update the method undef code in `.included` to handle already-undefined methods gracefully.

Allowing Dry::Configurable to be included multiple times across subclasses now allows for uses cases like a subclass adjusting various extension settings, e.g. `config_class`.

### CODE REVIEW

File: lib/dry/configurable/errors.rb [hunk 1]
```diff
@@ -7,7 +7,6 @@ module Dry
   module Configurable
     Error = Class.new(::StandardError)
 
-    AlreadyIncludedError = Class.new(Error)
```
Comment (flash-gordon) [inline lib/dry/configurable/errors.rb:10]: Hard to tell, but it may be a breaking change if somebody references this constant in their code

File: lib/dry/configurable/extension.rb [hunk 2]
```diff
@@ -36,8 +34,8 @@ def included(klass)
           prepend(Initializer)
```
Comment (flash-gordon) [inline lib/dry/configurable/extension.rb:34]: Now that these lines are reachable, things are more tricky. `include` and `prepend` work differently, `include` will add a module to the ancestors' chain only once, whereas `prepend` does it every time. We should keep this in mind

File: lib/dry/configurable/errors.rb (hunk 1 above)
Comment (timriley) [reply lib/dry/configurable/errors.rb:10]: Ah yes, thanks for noting that. I'll put the class back in for now, along with a comment that notes it will be deprecated soon.

Comment (flash-gordon) [reply lib/dry/configurable/errors.rb:10]: btw we have a mechanism for deprecating constants in dry-core

Comment (timriley) [reply lib/dry/configurable/errors.rb:10]: Thanks again @flash-gordon, have addressed this in ca63b93 :)

**Comment by timriley** [comment]
@flash-gordon FYI, this is the reason for this PR: [hanami/hanami@`86db090` (#1430)](https://github.com/hanami/hanami/pull/1430/commits/86db09019de42ee5a3fadc3334cbf866ed05183e)

`Dry::System::Provider::Source` already includes `Dry::Configurable`, providing a generic `config` object, which is fine in most circumstances. For the Hanami's `DB` provider source, however, we have fairly advanced config needs, so I want to provide a custom config API. By re-including `Dry::Configurable` and providing the `config_class:` option, we can now do that.

How does that feel to you?

I thought about adding a dedicated API to `Provider::Source` for specifying a custom config class, but that felt like overkill when I'm not sure how widely it will be needed, and our needs for the `DB` provider can be taken care of by a straightforward `include`. One downside of this approach is that it presumes knowledge of `Provider::Source`'s internals (i.e. that it uses `Dry::Configurable`), but given the close relationship between Hanami and dry-rb, I think we can manage this.

File: lib/dry/configurable/extension.rb (hunk 2 above)
Comment (timriley) [reply lib/dry/configurable/extension.rb:34]: Good call. I'll look to expand the testing to cover a few more possible edge cases before I mark this ready for a final review.

**Comment by timriley** [comment]
@flash-gordon I've expanded the tests to make sure nothing surprising happens with the multiple prepended `#initialize` methods. See the last commit above. It all looks good to me, things work as they should.

Given this, are you happy to see this be merged? Thanks again for checking!

**Review by flash-gordon — APPROVED** [review]
